File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/access/DeepLinkParser.kt
package mx.nexara.mobile.nativeapp.access

import android.net.Uri
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

sealed class DeepLinkDestination {
    data object Notifications : DeepLinkDestination()

    /** Módulo; [entityId] y [params] abren detalle cuando aplica. */
    data class Module(
        val panel: PanelId,
        val key: String,
        val entityId: Long? = null,
        val params: Map<String, String> = emptyMap(),
    ) : DeepLinkDestination()
}

/**
 * Traduce rutas web (`relatedUrl` de notificaciones, push, `nexara://`) a destinos
 * de la app. Espejo de `coreSurfaceRedirect` (`apps/web/lib/core-surface.ts`):
 *
 * - Core conserva: `/erp/actividades/:id[/evidencias|/historial]`,
 *   `/erp/pizarra[/:userId][?vista=]`, `/erp/mis-actividades`,
 *   `/erp/asistencias[?tab=comidas]`, `/erp/chat`, `/erp/my-profile`, `/erp/clientes`.
 * - Paneles fuera de /erp: el detalle y las evidencias de actividad conservan el
 *   id, «Mi perfil» y «Mis actividades» van a su equivalente, todo lo demás cae en
 *   la casa de Core (Actividades).
 * - El portal (`/tickets`, `/portal`) solo lo abre una cuenta de cliente o de
 *   sucursal; eso lo decide [PanelAccessResolver] al consumir el destino.
 */
object DeepLinkParser {

    /** Parámetro con el id de la persona en `/erp/pizarra/:userId`. */
    const val BOARD_USER_PARAM = "userid"

    /** `/erp/pizarra`: la casa de Core. */
    val CORE_HOME = DeepLinkDestination.Module(panel = PanelId.ERP, key = CoreKeys.ACTIVITIES)

    /** Rutas web tipo `/erp/actividades/123` o URL absoluta con path y query. */
    fun parseWebPath(pathOrUrl: String): DeepLinkDestination? {
        val trimmed = pathOrUrl.trim()
        if (trimmed.isBlank()) return null

        val (pathPart, queryPart) = when {
            trimmed.startsWith("http://") || trimmed.startsWith("https://") -> {
                val uri = Uri.parse(trimmed)
                Pair(uri.path?.trim('/') ?: "", uri.encodedQuery)
            }
            else -> {
                val qIdx = trimmed.indexOf('?')
                if (qIdx >= 0) {
                    Pair(trimmed.substring(0, qIdx).trim('/'), trimmed.substring(qIdx + 1))
                } else {
                    Pair(trimmed.trim('/'), null)
                }
            }
        }

        val params = parseQueryParams(queryPart)
        val segments = pathPart.split('/').filter { it.isNotBlank() }
        if (segments.isEmpty() && params.isEmpty()) return null
        return parseSegments(segments, params)
    }

    fun parse(uri: Uri?): DeepLinkDestination? {
        if (uri == null) return null
        val segments = buildList {
            val host = uri.host?.lowercase()
            val pathSegs = uri.pathSegments.orEmpty()
            if (!host.isNullOrBlank() && !host.contains(".")) {
                add(host)
                addAll(pathSegs)
            } else {
                addAll(pathSegs)
            }
        }.filter { it.isNotBlank() }

        val params = uri.queryParameterNames.orEmpty()
            .mapNotNull { name ->
                uri.getQueryParameter(name)?.let { name.lowercase() to it }
            }
            .toMap()

        if (segments.isEmpty() && params.isEmpty()) return null
        return parseSegments(segments, params)
    }

    private val ERP_HEADS = setOf("erp", "core")
    private val PORTAL_HEADS = setOf("portal", "tickets")

    /** `NON_ERP_PANEL_RE` de core-surface.ts. */
    private val NON_ERP_HEADS = setOf(
        "ops", "crm", "studio", "lab", "integra", "finance", "hr", "sales", "console",
        "consola", "contabilidad", "people", "operacion", "noc", "support", "ventas",
    )

    private val NOTIFICATION_SEGMENTS = setOf("notifications-center", "notifications", "notificaciones")

    private fun parseSegments(
        segments: List<String>,
        params: Map<String, String>,
    ): DeepLinkDestination? {
        if (segments.isEmpty()) return CORE_HOME
        val lower = segments.map { it.lowercase() }
        if (lower.last() in NOTIFICATION_SEGMENTS) return DeepLinkDestination.Notifications

        val head = lower.first()
        return when (head) {
            in PORTAL_HEADS -> portalDestination(lower.drop(1), params)
            in NON_ERP_HEADS -> legacyPanelDestination(head, lower.drop(1), params)
            in ERP_HEADS -> coreDestination(lower.drop(1), params)
            // `nexara://actividades/5`: sin prefijo de panel, se lee como Core.
            else -> coreDestination(lower, params)
        }
    }

    // ── Core (/erp) ─────────────────────────────────────────────────────────

    private val ACTIVITY_SEGMENTS = setOf("actividades", "activities", "pizarra")
    private val MY_ACTIVITIES_SEGMENTS = setOf("mis-actividades", "my-activities")
    private val EVIDENCE_LIST_SEGMENTS = setOf("my-evidences", "mis-evidencias", "evidences", "evidencias")
    private val ATTENDANCE_SEGMENTS = setOf("asistencias", "asistencia", "attendance")
    private val LUNCH_SEGMENTS = setOf("lunch-breaks", "my-lunch-breaks", "comidas")
    private val PROFILE_SEGMENTS = setOf("my-profile", "mi-perfil")
    private val CLIENT_SEGMENTS = setOf("clientes")

    // ── Módulos de «Más» (`CoreExtraModule`): la app abre su ficha, la web su página. ──
    private val VEHICLE_SEGMENTS = setOf("vehiculos", "vehicles", "mis-vehiculos", "my-vehicles")
    private val TOOL_SEGMENTS = setOf("herramientas", "tools")
    private val WAREHOUSE_SEGMENTS = setOf("almacen", "warehouse", "inventario")
    private val PROJECT_SEGMENTS = setOf("proyectos", "projects")
    private val ORGCHART_SEGMENTS = setOf("organigrama", "orgchart")

    /**
     * Viáticos. La web los tiene bajo `/erp/finance/viatics`, y los enlaces
     * viejos bajo `/ops/viatics` y `/ops/my-viatics`; las tres formas abren la
     * misma pantalla, conservando el id cuando viene.
     */
    private val VIATIC_SEGMENTS = setOf("viatics", "viaticos", "my-viatics", "mis-viaticos")

    private fun coreDestination(parts: List<String>, params: Map<String, String>): DeepLinkDestination {
        val extra = params.filterKeys { it !in ENTITY_ID_QUERY_KEYS }
        val first = parts.firstOrNull() ?: return CORE_HOME
        val second = parts.getOrNull(1)
        val activityId = params.longParam("activityid")

        return when {
            first == "pizarra" -> {
                val personId = second?.toLongOrNull()?.takeIf { it > 0L }
                module(
                    CoreKeys.ACTIVITIES,
                    params = when {
                        // Con persona se abre su día: ahí no hay pestañas que elegir.
                        personId != null -> extra + (BOARD_USER_PARAM to personId.toString())
                        // Quien pidió una vista concreta manda.
                        extra.containsKey("vista") -> extra
                        // «Pizarra» es el equipo. Sin esto el enlace caía en el
                        // valor por defecto de la pantalla, «Mis actividades», y
                        // un aviso que decía «ve a la pizarra» te dejaba viendo
                        // tu propia lista — la misma pantalla que
                        // `mis-actividades`, sin forma de notar la diferencia.
                        else -> extra + ("vista" to "equipo")
                    },
                )
            }
            first in MY_ACTIVITIES_SEGMENTS && activityId != null ->
                module(CoreKeys.ACTIVITIES, entityId = activityId, params = extra)
            first in MY_ACTIVITIES_SEGMENTS -> module(CoreKeys.MY_ACTIVITIES, params = extra)
            first in ACTIVITY_SEGMENTS -> {
                val id = second?.toLongOrNull()?.takeIf { it > 0L } ?: activityId
                if (id == null) return module(CoreKeys.ACTIVITIES, params = extra)
                val tab = parts.getOrNull(2)?.let { ACTIVITY_DETAIL_TABS[it] }
                module(
                    CoreKeys.ACTIVITIES,
                    entityId = id,
                    params = if (tab != null) extra + ("tab" to tab) else extra,
                )
            }
            first in EVIDENCE_LIST_SEGMENTS && activityId != null ->
                module(CoreKeys.ACTIVITIES, entityId = activityId, params = extra + ("tab" to "evidencias"))
            // KPIs del equipo antes que Asistencias: viven dentro de la misma ruta.
            first in ATTENDANCE_SEGMENTS && second == "indicadores" ->
                module(CoreKeys.KPIS_EQUIPO, params = extra)
            first in ATTENDANCE_SEGMENTS -> module(CoreKeys.ATTENDANCE, params = extra)
            first == "cotizaciones" ->
                module(CoreKeys.COTIZACIONES, entityId = second?.toLongOrNull()?.takeIf { it > 0L }, params = extra)
            first in PROJECT_SEGMENTS ->
                module(CoreKeys.PROYECTOS, entityId = second?.toLongOrNull()?.takeIf { it > 0L }, params = extra)
            first in WAREHOUSE_SEGMENTS && second in TOOL_SEGMENTS -> module(CoreKeys.HERRAMIENTAS, params = extra)
            first in WAREHOUSE_SEGMENTS -> module(CoreKeys.ALMACEN, params = extra)
            first in TOOL_SEGMENTS -> module(CoreKeys.HERRAMIENTAS, params = extra)
            first in VEHICLE_SEGMENTS ->
                module(CoreKeys.VEHICULOS, entityId = second?.toLongOrNull()?.takeIf { it > 0L }, params = extra)
            // `/erp/finance/viatics[/:id]` — donde la web tiene viáticos.
            first == "finance" && second in VIATIC_SEGMENTS ->
                module(
                    CoreKeys.VIATICOS,
                    entityId = parts.getOrNull(2)?.toLongOrNull()?.takeIf { it > 0L },
                    params = extra,
                )
            first in VIATIC_SEGMENTS ->
                module(CoreKeys.VIATICOS, entityId = second?.toLongOrNull()?.takeIf { it > 0L }, params = extra)
            first in ORGCHART_SEGMENTS || (first == "hr" && second in ORGCHART_SEGMENTS) ->
                module(CoreKeys.ORGANIGRAMA, params = extra)
            // `/erp/hr/attendance` y `/erp/hr/lunch-breaks` (appUrls viejos) → Asistencias.
            first == "hr" && second in ATTENDANCE_SEGMENTS -> module(CoreKeys.ATTENDANCE, params = extra)
            first in LUNCH_SEGMENTS || (first == "hr" && second in LUNCH_SEGMENTS) ->
                module(CoreKeys.ATTENDANCE, params = extra + ("tab" to "comidas"))
            first == CoreKeys.CHAT -> module(
                CoreKeys.CHAT,
                entityId = second?.toLongOrNull()?.takeIf { it > 0L } ?: params.longParam("channel"),
                params = extra,
            )
            first in PROFILE_SEGMENTS -> module(CoreKeys.MY_PROFILE, params = extra)
            first in CLIENT_SEGMENTS -> module(
                CoreKeys.CLIENTS,
                entityId = second?.toLongOrNull()?.takeIf { it > 0L },
                params = extra,
            )
            else -> CORE_HOME
        }
    }

    // ── Paneles fuera de /erp (coreSurfaceRedirect) ─────────────────────────

    private val LEGACY_ACTIVITY_HEADS = setOf("ops", "operacion")

    private fun legacyPanelDestination(
        head: String,
        parts: List<String>,
        params: Map<String, String>,
    ): DeepLinkDestination {
        val extra = params.filterKeys { it !in ENTITY_ID_QUERY_KEYS }
        val first = parts.firstOrNull()

        if (head in LEGACY_ACTIVITY_HEADS && first in setOf("activities", "actividades")) {
            val id = parts.getOrNull(1)?.takeIf { s -> s.all(Char::isDigit) }?.toLongOrNull()
            if (id != null && id > 0L) {
                val suffix = parts.getOrNull(2)
                return if (suffix == "evidences" || suffix == "evidencias") {
                    module(CoreKeys.ACTIVITIES, entityId = id, params = extra + ("tab" to "evidencias"))
                } else {
                    module(CoreKeys.ACTIVITIES, entityId = id, params = extra)
                }
            }
        }

        val activityId = params.longParam("activityid")
        if (head in LEGACY_ACTIVITY_HEADS && parts.size == 1 && activityId != null) {
            if (first in EVIDENCE_LIST_SEGMENTS) {
                return module(CoreKeys.ACTIVITIES, entityId = activityId, params = extra + ("tab" to "evidencias"))
            }
            if (first in MY_ACTIVITIES_SEGMENTS || first in setOf("activities", "actividades")) {
                return module(CoreKeys.ACTIVITIES, entityId = activityId, params = extra)
            }
        }

        if (parts.lastOrNull() == "my-profile") return module(CoreKeys.MY_PROFILE)
        if (head in LEGACY_ACTIVITY_HEADS && parts.size == 1 && first in MY_ACTIVITIES_SEGMENTS) {
            return module(CoreKeys.MY_ACTIVITIES)
        }
        // Viáticos vive en `/erp/finance/viatics`, y «finance» es un panel viejo:
        // `/finance/viatics/:id` y `/ops/viatics/:id` abren la misma pantalla.
        if (first in VIATIC_SEGMENTS && (head == "ops" || head == "finance")) {
            return module(
                CoreKeys.VIATICOS,
                entityId = parts.getOrNull(1)?.toLongOrNull()?.takeIf { it > 0L }
                    ?: params.longParam("highlight"),
                params = extra,
            )
        }

        // Módulos de OPS que se mudaron a Core (`coreSurfaceRedirect` → `MOVED_TO_CORE` en la web):
        // vehículos, herramientas y proyectos abren su ficha de «Más».
        if (head == "ops") {
            when {
                first in VEHICLE_SEGMENTS -> return module(
                    CoreKeys.VEHICULOS,
                    entityId = parts.getOrNull(1)?.toLongOrNull()?.takeIf { it > 0L },
                    params = extra,
                )
                first in TOOL_SEGMENTS -> return module(CoreKeys.HERRAMIENTAS, params = extra)
                first in PROJECT_SEGMENTS -> return module(
                    CoreKeys.PROYECTOS,
                    entityId = parts.getOrNull(1)?.toLongOrNull()?.takeIf { it > 0L },
                    params = extra,
                )
            }
        }
        return CORE_HOME
    }

    // ── Portal de clientes ──────────────────────────────────────────────────

    private val PORTAL_KEY_ALIASES = mapOf(
        "sucursales" to "branches",
        "solicitudes" to "requests",
        "inventarios" to "inventories",
        "feedback" to "feedback-pending",
        "mi-perfil" to "profile",
        "my-profile" to "profile",
    )

    private fun portalDestination(parts: List<String>, params: Map<String, String>): DeepLinkDestination {
        val pathId = parts.lastOrNull()?.takeIf { it.all(Char::isDigit) }?.toLongOrNull()
        val keyParts = if (pathId != null) parts.dropLast(1) else parts
        val raw = keyParts.lastOrNull() ?: "portal"
        val key = PORTAL_KEY_ALIASES[raw] ?: raw
        return DeepLinkDestination.Module(
            panel = PanelId.PORTAL,
            key = key,
            entityId = pathId ?: params.longParam("highlight") ?: params.longParam("id"),
            params = params.filterKeys { it !in ENTITY_ID_QUERY_KEYS },
        )
    }

    // ── Utilidades ──────────────────────────────────────────────────────────

    private val ENTITY_ID_QUERY_KEYS = setOf("highlight", "id", "channel", "activityid")

    /**
     * Sufijo de `/erp/actividades/{id}/<sufijo>` → clave de pestaña del detalle.
     * La web conserva `evidencias` e `historial`; los sufijos en inglés son de
     * los enlaces viejos de OPS.
     */
    private val ACTIVITY_DETAIL_TABS = mapOf(
        "evidencias" to "evidencias",
        "evidences" to "evidencias",
        "historial" to "historial",
        "history" to "historial",
    )

    private fun module(
        key: String,
        entityId: Long? = null,
        params: Map<String, String> = emptyMap(),
    ) = DeepLinkDestination.Module(panel = PanelId.ERP, key = key, entityId = entityId, params = params)

    private fun parseQueryParams(query: String?): Map<String, String> {
        if (query.isNullOrBlank()) return emptyMap()
        return query.split('&')
            .mapNotNull { pair ->
                val idx = pair.indexOf('=')
                if (idx <= 0) return@mapNotNull null
                val name = URLDecoder.decode(pair.substring(0, idx), StandardCharsets.UTF_8).lowercase()
                val value = URLDecoder.decode(pair.substring(idx + 1), StandardCharsets.UTF_8)
                name to value
            }
            .toMap()
    }

    private fun Map<String, String>.longParam(name: String): Long? =
        this[name]?.toLongOrNull()?.takeIf { it > 0L }
}

File: apps/mobile-native/ios/NexaraApp/Access/DeepLinkParser.swift
import Foundation

/// Pantalla de Core a la que lleva un enlace.
struct CoreLink: Equatable {
    var module: CoreModule
    /// Detalle de actividad (`/erp/actividades/:id`).
    var activityId: Int? = nil
    /// Actividades: `detalle` / `evidencias` / `historial`. Asistencias: `comidas`.
    var tab: String? = nil
    /// Día de una persona (`/erp/pizarra/:userId`).
    var boardUserId: Int? = nil
    /// Vista de Actividades (`mias` / `equipo`).
    var vista: String? = nil
    var chatChannelId: Int64? = nil
    var chatMessageId: Int64? = nil
    /// Cliente (`/erp/clientes/:id`).
    var entityId: Int64? = nil
    /// Módulo del hub «Más» (Almacén, Vehículos…). Con él, `module` se queda en
    /// `.actividades`: el shell lo abre encima de la pestaña actual.
    var extra: CoreExtraModule? = nil

    /// `CORE_HOME_PATH` = `/erp/pizarra`.
    static let home = CoreLink(module: .actividades)
}

/// Destino de un deep link (`nexara://`, URL https o ruta web relativa).
enum DeepLinkDestination: Equatable {
    case notifications
    case core(CoreLink)
    /// Portal externo de clientes (`/tickets/...`, «NO se toca» en la web).
    case portal(key: String, entityId: Int64?)
}

/// Solo existe ERP (Core): todo lo demás cae en Actividades, igual que
/// `coreSurfaceRedirect` en `apps/web/lib/core-surface.ts`.
enum DeepLinkParser {
    /// `NON_ERP_PANEL_RE` de la web.
    private static let nonErpHeads: Set<String> = [
        "ops", "crm", "studio", "lab", "integra", "finance", "hr", "sales", "console", "consola",
        "contabilidad", "people", "operacion", "noc", "support", "ventas",
    ]
    private static let portalHeads: Set<String> = ["tickets", "portal"]
    private static let activityHeads: Set<String> = ["activities", "actividades"]
    private static let myActivityHeads: Set<String> = ["my-activities", "mis-actividades"]
    private static let evidenceHeads: Set<String> = ["evidences", "evidencias", "my-evidences", "mis-evidencias"]

    static func parse(_ url: URL) -> DeepLinkDestination? {
        var segments = url.pathComponents.filter { $0 != "/" }
        let scheme = (url.scheme ?? "").lowercased()
        // En `nexara://erp/actividades/12` el primer segmento llega como host.
        if scheme != "http", scheme != "https", let host = url.host, !host.isEmpty, !host.contains(".") {
            segments.insert(host, at: 0)
        }
        var params: [String: String] = [:]
        if let components = URLComponents(url: url, resolvingAgainstBaseURL: false) {
            for item in components.queryItems ?? [] {
                guard let value = item.value else { continue }
                params[item.name] = value
                params[item.name.lowercased()] = value
            }
        }
        return parse(segments: segments, params: params)
    }

    /// Ruta web (`/erp/actividades/12/evidencias`) o URL absoluta, como llega en
    /// `relatedUrl` de las notificaciones.
    static func parseWebPath(_ pathOrUrl: String) -> DeepLinkDestination? {
        let trimmed = pathOrUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let lower = trimmed.lowercased()
        if lower.hasPrefix("http://") || lower.hasPrefix("https://") || lower.hasPrefix("nexara://") {
            guard let url = URL(string: trimmed)
                ?? URL(string: trimmed.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "") else {
                return nil
            }
            return parse(url)
        }
        let path = trimmed.hasPrefix("/") ? trimmed : "/" + trimmed
        let base = "https://app.nexara.local"
        guard let url = URL(string: base + path)
            ?? URL(string: base + (path.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "")) else {
            return nil
        }
        return parse(url)
    }

    private static func positiveInt(_ raw: String?) -> Int? {
        guard let raw, let value = Int(raw), value > 0 else { return nil }
        return value
    }

    private static func positiveInt64(_ raw: String?) -> Int64? {
        guard let raw, let value = Int64(raw), value > 0 else { return nil }
        return value
    }

    /// Enlace a un módulo del hub «Más».
    private static func extraLink(_ module: CoreExtraModule) -> CoreLink {
        CoreLink(module: .actividades, extra: module)
    }

    private static func parse(segments rawSegments: [String], params: [String: String]) -> DeepLinkDestination? {
        let segments = rawSegments.map { $0.lowercased() }.filter { !$0.isEmpty }
        guard let head = segments.first else { return nil }

        if segments.contains("notifications-center") || segments == ["notifications"] {
            return .notifications
        }
        if portalHeads.contains(head) {
            let rest = Array(segments.dropFirst())
            let entityId = positiveInt64(rest.last)
            let parts = entityId != nil ? Array(rest.dropLast()) : rest
            return .portal(key: parts.last ?? "home", entityId: entityId)
        }
        if nonErpHeads.contains(head) {
            return .core(redirectNonErp(segments: segments, params: params))
        }
        let parts = head == "erp" ? Array(segments.dropFirst()) : segments
        return erpDestination(parts: parts, params: params)
    }

    /// `coreSurfaceRedirect`: detalle y evidencias de actividad conservan el id,
    /// «Mi perfil» y «Mis actividades» van a su equivalente; lo demás, a casa.
    private static func redirectNonErp(segments: [String], params: [String: String]) -> CoreLink {
        let head = segments[0]
        let rest = Array(segments.dropFirst())

        if head == "ops", rest.count >= 2, activityHeads.contains(rest[0]), let id = positiveInt(rest[1]) {
            let suffix = rest.count >= 3 ? rest[2] : ""
            let tab = (suffix == "evidences" || suffix == "evidencias") ? "evidencias" : "detalle"
            return CoreLink(module: .actividades, activityId: id, tab: tab)
        }

        if head == "ops", rest.count == 1, let activityId = positiveInt(params["activityId"] ?? params["activityid"]) {
            if evidenceHeads.contains(rest[0]) {
                return CoreLink(module: .actividades, activityId: activityId, tab: "evidencias")
            }
            if activityHeads.contains(rest[0]) || myActivityHeads.contains(rest[0]) {
                return CoreLink(module: .actividades, activityId: activityId, tab: "detalle")
            }
        }

        if segments.last == "my-profile" {
            return CoreLink(module: .perfil)
        }
        if head == "ops", rest.count == 1, myActivityHeads.contains(rest[0]) {
            return CoreLink(module: .actividades, vista: "mias")
        }
        // Vehículos, herramientas y proyectos de Operaciones viven ahora en
        // /erp (la web ya manda `/ops/projects` a `/erp/proyectos`).
        if head == "ops", let first = rest.first {
            switch first {
            case "vehicles", "vehiculos", "my-vehicles", "mis-vehiculos":
                return extraLink(.vehiculos)
            case "tools", "herramientas":
                return extraLink(.herramientas)
            case "projects", "proyectos":
                return extraLink(.proyectos)
            default:
                break
            }
        }
        return CoreLink.home
    }

    /// Rutas de `/erp` que existen en Core; cualquier otra cae en Actividades.
    private static func erpDestination(parts: [String], params: [String: String]) -> DeepLinkDestination {
        guard let first = parts.first else { return .core(CoreLink.home) }
        let second: String? = parts.count >= 2 ? parts[1] : nil
        let queryActivityId = positiveInt(params["activityId"] ?? params["activityid"])

        if activityHeads.contains(first) || myActivityHeads.contains(first) {
            if parts.count >= 2, let id = positiveInt(parts[1]) {
                let suffix = parts.count >= 3 ? parts[2] : ""
                let tab: String
                switch suffix {
                case "evidencias", "evidences": tab = "evidencias"
                case "historial", "history": tab = "historial"
                default: tab = "detalle"
                }
                return .core(CoreLink(module: .actividades, activityId: id, tab: tab))
            }
            if let queryActivityId {
                return .core(CoreLink(module: .actividades, activityId: queryActivityId, tab: "detalle"))
            }
            let vista = myActivityHeads.contains(first) ? "mias" : params["vista"]
            return .core(CoreLink(module: .actividades, vista: vista))
        }

        switch first {
        case "pizarra":
            var link = CoreLink(module: .actividades, vista: params["vista"])
            if parts.count >= 2, let userId = positiveInt(parts[1]) {
                link.boardUserId = userId
                link.vista = "equipo"
            }
            return .core(link)
        case "asistencias", "asistencia", "attendance":
            // KPIs del equipo cuelga de Asistencias en la web, pero es otro módulo.
            if second == "indicadores" {
                return .core(extraLink(.kpisEquipo))
            }
            return .core(CoreLink(module: .asistencias, tab: params["tab"]))
        case "almacen", "warehouse":
            return .core(extraLink(second == "herramientas" ? .herramientas : .almacen))
        case "vehiculos":
            // Incluye `mis-vehiculos`, `:id` y `gps`.
            return .core(extraLink(.vehiculos))
        case "organigrama":
            return .core(extraLink(.organigrama))
        case "hr":
            // `/erp/hr/orgchart`; lo demás de RH no existe en Core.
            return .core(second == "orgchart" ? extraLink(.organigrama) : CoreLink.home)
        case "cotizaciones":
            return .core(extraLink(.cotizaciones))
        case "proyectos":
            return .core(extraLink(.proyectos))
        case "chat":
            let channel = parts.count >= 2 ? positiveInt64(parts[1]) : nil
            return .core(CoreLink(
                module: .chat,
                chatChannelId: channel ?? positiveInt64(params["channel"] ?? params["channelid"]),
                chatMessageId: positiveInt64(params["msg"] ?? params["messageid"])
            ))
        case "my-profile", "mi-perfil":
            return .core(CoreLink(module: .perfil))
        case "clientes":
            return .core(CoreLink(
                module: .clientes,
                vista: params["sector"],
                entityId: parts.count >= 2 ? positiveInt64(parts[1]) : nil
            ))
        case "notificaciones":
            return .notifications
        default:
            return .core(CoreLink.home)
        }
    }
}
